SARIF reporter

xmo-odoo opened the pull request:
|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

SARIF is a unified file format used to exchange information between static analysis tools (like pylint) and various types of formatters, meta-runners, broadcasters / alert system, ...

This implementation is ad-hoc, and non-validating.

Spec v Github
-------------

Turns out Github both doesn't implement all of SARIF (which makes sense) *and* requires a bunch of properties which the spec considers optional. The [official SARIF validator][] (linked to by both oasis and github) was used to validate the output of the reporter, ensuring that all the github requirements it flags are fulfilled, and fixing *some* of the validator's pet issues.

As of now the following issues are left unaddressed:

- azure requires `run.automationDetails`, looking at the spec I don't think it makes sense for the reporter to inject that, it's more up to the CI
- the validator wants a `run.versionControlProvenance`, same as above
- the validator wants rule names in PascalCase, lol
- the validator wants templated result messages, but without pylint providing the args as part of the `Message` that's a bit of a chore
- the validator wants `region` to include a snippet (the flagged content)
- the validator wants `physicalLocation` to have a `contextRegion` (most likely with a snippet)

On URIs
-------

The reporter makes use of URIs for artifacts (~files). Per ["guidance on the use of artifactLocation objects"][3.4.7], `uri` *should* capture the deterministic part of the artifact location and `uriBaseId` *should* capture the non-deterministic part. However as far as I can tell pylint has no requirement (and no clean way to require) consistent resolution roots: `path` is just relative to the cwd, and there is no requirement to have project-level files to use pylint. This makes the use of relative uris dodgy, but absolute uris are pretty much always broken for the purpose of *interchange* so they're not really any better.

As a side-note, Github [asserts][relative-uri-guidance]

> While this [nb: `originalUriBaseIds`] is not required by GitHub for the code scanning results to be displayed correctly, it is required to produce a valid SARIF output when using relative URI references.

However per [3.4.4][] this is incorrect, the `uriBaseId` can be resolved through end-user configuration, `originalUriBaseIds`, external information (e.g. envvars), or heuristics.

It would be nice to document the "relative root" via `originalUriBaseIds` (which may be omitted for that purpose per [3.14.14][], but per the above claiming a consistent project root is dodgy.

We *could* resolve known project files (e.g. pyproject.toml, tox.ini, etc...) in order to find a consistent root (project root, repo root, ...) and set / use that for relative URIs but that's a lot of additional complexity which I'm not sure is warranted at least for a first version.

Fixes #5493

[3.4.4]: https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10540869 
[3.4.7]:
https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10540872 
[3.14.14]: https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10540936
[relative-uri-guidance]:
https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#relative-uri-guidance-for-sarif-producers
[official SARIF validator]: https://sarifweb.azurewebsites.net/

xmo-odoo committed:
SARIF reporter

SARIF is a unified file format used to exchange information between
static analysis tools (like pylint) and various types of formatters,
meta-runners, broadcasters / alert system, ...

This implementation is ad-hoc, and non-validating.

Spec v Github
-------------

Turns out Github both doesn't implement all of SARIF (which makes
sense) and requires a bunch of properties which the spec considers
optional. The [official SARIF validator][] (linked to by both oasis
and github) was used to validate the output of the reporter, ensuring
that all the github requirements it flags are fulfilled, and
fixing *some* of the validator's pet issues.

As of now the following issues are left unaddressed:

- azure requires `run.automationDetails`, looking at the spec I don't
  think it makes sense for the reporter to inject that, it's more up
  to the CI
- the validator wants a `run.versionControlProvenance`, same as above
- the validator wants rule names in PascalCase, lol
- the validator wants templated result messages, but without pylint
  providing the args as part of the `Message` that's a bit of a chore
- the validator wants `region` to include a snippet (the flagged
  content)
- the validator wants `physicalLocation` to have a `contextRegion`
  (most likely with a snippet)

On URIs
-------

The reporter makes use of URIs for artifacts (~files). Per ["guidance
on the use of artifactLocation objects"][3.4.7], `uri` *should*
capture the deterministic part of the artifact location and
`uriBaseId` *should* capture the non-deterministic part. However as
far as I can tell pylint has no requirement (and no clean way to
require) consistent resolution roots: `path` is just relative to the
cwd, and there is no requirement to have project-level files to use
pylint. This makes the use of relative uris dodgy, but absolute uris
are pretty much always broken for the purpose of *interchange* so
they're not really any better.

As a side-note, Github [asserts][relative-uri-guidance]

> While this [nb: `originalUriBaseIds`] is not required by GitHub for
> the code scanning results to be displayed correctly, it is required
> to produce a valid SARIF output when using relative URI references.

However per [3.4.4][] this is incorrect, the `uriBaseId` can be
resolved through end-user configuration, `originalUriBaseIds`,
external information (e.g. envvars), or heuristics.

It would be nice to document the "relative root" via
`originalUriBaseIds` (which may be omitted for that purpose per
[3.14.14][], but per the above claiming a consistent project root is
dodgy.

We *could* resolve known project files (e.g. pyproject.toml, tox.ini,
etc...) in order to find a consistent root (project root, repo root,
...) and set / use that for relative URIs but that's a lot of
additional complexity which I'm not sure is warranted at least for a
first version.

Fixes #5493

[3.4.4]: https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10540869
[3.4.7]:
https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10540872
[3.14.14]: https://docs.oasis-open.org/sarif/sarif/v2.1.0/csprd01/sarif-v2.1.0-csprd01.html#_Toc10540936
[relative-uri-guidance]:
https://docs.github.com/en/code-security/code-scanning/integrating-with-code-scanning/sarif-support-for-code-scanning#relative-uri-guidance-for-sarif-producers
[official SARIF validator]: https://sarifweb.azurewebsites.net/

Pierre-Sassoulas commented on pylint/reporters/json_reporter.py:
```diff
@@ -196,6 +200,120 @@ def serialize_stats(self) -> dict[str, str | int | dict[str, int]]:
         }
 
 
+class SARIFReporter(BaseReporter):
```
Should be in ``pylint/reporters/sarif_reporter.py`` not ``pylint/reporters/json_reporter.py``

Pierre-Sassoulas commented on pylint/reporters/sarif_types.py:
```diff
@@ -0,0 +1,134 @@
+"""
```
We can create a package and move this to ``pylint/reporters/sarif/types.py`` I guess

Pierre-Sassoulas reviewed (changes requested):
Thank you, this is great. 

I'm not a SARIF expert so I have some code placement comments, and I hope someone more knowledgeable in SARIF will have an opinion about the actual content.

nvuillam commented:
As long as the sarif validator validates the SARIF output, it's good :)

Pierre-Sassoulas commented on pylint/reporters/sarif_reporter.py:
```diff
@@ -0,0 +1,260 @@
+                            "fullName": f"pylint {pylint.__version__}",
+                            "version": pylint.__version__,
+                            # should be versioned but not all versions are kept so...
+                            "informationUri": "https://pylint.readthedocs.io/",
```
Yeah latest is good enough imo.

Pierre-Sassoulas reviewed (commented):
Implementation LGTM, it seems we don't need to cover everything as long as we add sarif validation in the test suite ?

xmo-odoo committed:
Unpin pre-commit version in tox

The config uses the `pre-commit/pre-commit-hook` repository which
since v5.0 has multiple hooks using the `pre-commit` stage which was
only added in pre-commit [3.2.0][] causing a run failure when run
under a pre-commit pinned to 2.2. From a very surface looks it seems
like https://pre-commit.ci/ is anal about keeping everything on latest
so pinning pre-commit in tox seems like it can only cause problems (by
drifting behind the actual CI).

Of note: it looks like the `minimum_pre_commit_version` stricture on
hooks does not do anything: all the `pre-commit` hooks are filtered on
version except [`destroyed-symlinks`][] but I get an error on
`check-added-large-file`:

    ==> At Hook(id='check-added-large-files')
    ==> At key: stages
    ==> At index 0
    =====> Expected one of commit, commit-msg, manual, merge-commit, post-checkout, post-commit, post-merge, post-rewrite, prepare-commit-msg, push but got: 'pre-commit'

[3.2.0]: https://github.com/pre-commit/pre-commit/blob/main/CHANGELOG.md#320---2023-03-17
[`destroyed-symlinks`]: https://github.com/pre-commit/pre-commit-hooks/commit/c7d1e85f24382186efd2c67b95891b00dc3c4505

Pierre-Sassoulas commented on tests/reporters/unittest_sarif_reporter.py:
```diff
@@ -0,0 +1,162 @@
+                f"Unable to retrieve schema v{SCHEMA_VERSION}: {res.reason}"
+            )
+        schema = res.text
+        pytestconfig.cache.set(CACHE_KEY, schema)
```
Let's create the repository externally and transfer it to pylint-dev once it's ready then.

Pierre-Sassoulas replied:
There's an example here : https://github.com/pylint-dev/astroid/issues/1312#issuecomment-3795726065, the functional test remarks does not apply for the sarif reporter as what need to be checked is not the behavior of pylint on some code, but the reporter itself.

Pierre-Sassoulas commented:
@xmo-odoo sorry for dropping the ball here. The pylint-sarif repo at https://github.com/xmo-odoo/pylint-sarif LGTM. We can transfer it to pylint-dev and give you admin right on it. Then we still need to release it it seems ? And this PR become simply adding an ad for it in pylint's documentation :)
